deps/happy-eyeballs/happy-eyeballs.c

```diff
@@ -0,0 +1,528 @@
+		// we're not last...
+		free(args);
+		return NULL;
+	}
+	// Ok, we are last. Coalesce errors and signal completion.
+	context->error_message = "Failed to connect";
+	os_event_signal(context->race_completed_event);
+	free(args);
+	return NULL;
+}
+
+static int launch_worker(struct happy_eyeballs_ctx *context, int index,
+			 struct addrinfo *addr)
+{
+#ifdef _WIN32
+	SOCKET fd = WSASocket(addr->ai_family, SOCK_STREAM, IPPROTO_TCP, NULL,
+			      0, WSA_FLAG_OVERLAPPED);
+#else
+	SOCKET fd = socket(addr->ai_family, SOCK_STREAM, IPPROTO_TCP);
+#endif
+	struct happy_eyeballs_candidate *candidate =
+		&context->candidates[index];
+	candidate->sockfd = fd;
+	struct happy_connect_woker_args *args =
+		(struct happy_connect_woker_args *)malloc(
+			sizeof(struct happy_connect_woker_args));
+	if (args == NULL) {
+		context->error = ENOMEM;
+		context->error_message =
+			"happy-eyeballs: Failed to allocate memory for worker context";
+		return STATUS_FAILURE;
+	}
+	args->sockfd = (int)fd;
+	args->address = addr;
+	args->context = context;
+	args->candidate = candidate;
+
+	int result = os_event_init(&candidate->socket_completed_event,
+				   OS_EVENT_TYPE_MANUAL);
+	if (result != 0) {
+		// failure to create the
+		context->error = result;
+		context->error_message =
+			"happy-eyeballs: Failed to initialize socket completed event";
+		return STATUS_FAILURE;
+	}
+
+	result = pthread_create(&candidate->thread, NULL, happy_connect_worker,
+				args);
+	if (result != 0) {
+		// failure to start the worker thread
+		context->error = result;
+		context->error_message = strerror(result);
+		os_event_destroy(candidate->socket_completed_event);
+		return STATUS_FAILURE;
+	}
+
+	candidate->valid = true;
+	return STATUS_SUCCESS;
+}
+
+int happy_eyeballs_init(struct happy_eyeballs_ctx **context)
+{
+	if (context == NULL) {
+		return STATUS_INVALID_ARGUMENT;
+	}
+
+	*context = (struct happy_eyeballs_ctx *)malloc(
+		sizeof(struct happy_eyeballs_ctx));
+	if (*context == NULL) {
+		return STATUS_FAILURE;
+	}
+
+	memset(*context, 0, sizeof(struct happy_eyeballs_ctx));
+
+	// race_completed_event will be signalled when there is a winner
+	int result = os_event_init(&(*context)->race_completed_event,
+				   OS_EVENT_TYPE_MANUAL);
+	if (result != 0) {
+		(*context)->error = result;
+		(*context)->error_message =
+			"happy-eyeballs: Failed to initialize race completed event";
+		return STATUS_FAILURE;
+	}
+
+	// this mutex is used to avoid the situation where we may have two simultaneous winners and inconsistent
+	// values set to the context.
+	result = pthread_mutex_init(&(*context)->winner_mutex, NULL);
+	if (result != 0) {
+		(*context)->error = result;
+		(*context)->error_message =
+			"happy-eyeballs: Failed to initialize winner mutex";
+		return STATUS_FAILURE;
+	}
+	return STATUS_SUCCESS;
+}
+
+int happy_eyeballs_connect(struct happy_eyeballs_ctx *context,
+			   const char *hostname, int port)
+{
+	if (hostname == NULL || context == NULL || port == 0) {
```

Comment:
I won't mention all of them, but OBS prefers typically no curlies on if statements with a single line